protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+# Allowing Non-Determinism in `ProcessProposal` (a.k.a. addressing #1174)
+
+## Context
+
+### Consensus Properties
+
+Byzantine Fault Tolerant (BFT) Consensus is usually specified by the following properties (2 invariants and the liveness property termination):
+
+- _agreement_: no two correct processes decide differently.
+- _validity_: the decided block `v` satisfies the predefined predicate `valid(v)`.
+- _termination_: all correct processes eventually decide.
+
+The consensus algorithm implemented in CometBFT (Tendermint) guarantees these properties for each height.
+
+> To be a bit more formal, valid it a mathematical function here that depends only on its inputs.
+> The arxiv paper glosses over some details here. In fact, valid is a function of the proposed block `v`
```

Comment:
A simplified and more generic version of this is to define the predicate as `valid(v, h)`, where `h` is the current height.

If the application adopting consensus is a blockchain, as it is, to each height `h` we have associated a deterministic state `s(h)`. Plus, we have the relation `s(h) = f(s(h-1), v(h-1))`, where `v(h-1)` is the consensus output of instance `h-1`. The point is that those details might be omitted when considering the _consensus_ protocol itself, as they are "application"-specific, here the application being the blockchain.